File: packages/frontend/apps/ios/App/App/Plugins/Preview/PreviewPlugin.swift
import Foundation
import Capacitor

private func resolveLocalFontDir(from fontURL: String) -> String? {
  let path: String
  if fontURL.hasPrefix("file://") {
    guard let url = URL(string: fontURL), url.isFileURL else {
      return nil
    }
    path = url.path
  } else {
    let candidate = (fontURL as NSString).standardizingPath
    guard candidate.hasPrefix("/") else {
      return nil
    }
    path = candidate
  }

  var isDirectory: ObjCBool = false
  if FileManager.default.fileExists(atPath: path, isDirectory: &isDirectory),
     isDirectory.boolValue
  {
    return path
  }

  let directory = (path as NSString).deletingLastPathComponent
  return directory.isEmpty ? nil : directory
}

private func resolveTypstFontDirs(from options: [AnyHashable: Any]?) throws -> [String]? {
  guard let rawFontUrls = options?["fontUrls"] else {
    return nil
  }

  guard let fontUrls = rawFontUrls as? [Any] else {
    throw NSError(
      domain: "PreviewPlugin",
      code: 1,
      userInfo: [
        NSLocalizedDescriptionKey: "Typst preview fontUrls must be an array of strings."
      ]
    )
  }

  var seenFontDirs = Set<String>()
  var orderedFontDirs = [String]()
  orderedFontDirs.reserveCapacity(fontUrls.count)

  for fontUrl in fontUrls {
    guard let fontURL = fontUrl as? String else {
      throw NSError(
        domain: "PreviewPlugin",
        code: 1,
        userInfo: [
          NSLocalizedDescriptionKey: "Typst preview fontUrls must be strings."
        ]
      )
    }

    guard let fontDir = resolveLocalFontDir(from: fontURL) else {
      throw NSError(
        domain: "PreviewPlugin",
        code: 1,
        userInfo: [
          NSLocalizedDescriptionKey: "Typst preview on mobile only supports local font file URLs or absolute font directories."
        ]
      )
    }

    if seenFontDirs.insert(fontDir).inserted {
      orderedFontDirs.append(fontDir)
    }
  }

  return orderedFontDirs
}

@objc(PreviewPlugin)
public class PreviewPlugin: CAPPlugin, CAPBridgedPlugin {
  public let identifier = "PreviewPlugin"
  public let jsName = "Preview"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "renderMermaidSvg", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "renderTypstSvg", returnType: CAPPluginReturnPromise),
  ]

  @objc func renderMermaidSvg(_ call: CAPPluginCall) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let code = try call.getStringEnsure("code")
        let options = call.getObject("options")
        let svg = try renderMermaidPreviewSvg(
          code: code,
          theme: options?["theme"] as? String,
          fontFamily: options?["fontFamily"] as? String,
          fontSize: (options?["fontSize"] as? NSNumber)?.doubleValue
        )
        call.resolve(["svg": svg])
      } catch {
        call.reject("Failed to render Mermaid preview, \(error)", nil, error)
      }
    }
  }

  @objc func renderTypstSvg(_ call: CAPPluginCall) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        let code = try call.getStringEnsure("code")
        let options = call.getObject("options")
        let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?.path
        let fontDirs = try resolveTypstFontDirs(from: options)
        let svg = try renderTypstPreviewSvg(code: code, fontDirs: fontDirs, cacheDir: cacheDir)
        call.resolve(["svg": svg])
      } catch {
        call.reject("Failed to render Typst preview, \(error)", nil, error)
      }
    }
  }
}

File: packages/frontend/apps/android/App/app/src/main/java/app/affine/pro/plugin/PreviewPlugin.kt
package app.affine.pro.plugin

import android.net.Uri
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import kotlinx.coroutines.Dispatchers
import timber.log.Timber
import uniffi.affine_mobile_native.renderMermaidPreviewSvg
import uniffi.affine_mobile_native.renderTypstPreviewSvg
import java.io.File

private fun JSObject.getOptionalString(key: String): String? {
    return if (has(key) && !isNull(key)) getString(key) else null
}

private fun JSObject.getOptionalDouble(key: String): Double? {
    return if (has(key) && !isNull(key)) getDouble(key) else null
}

private fun resolveLocalFontDir(fontUrl: String): String? {
    val uri = Uri.parse(fontUrl)
    val path = when {
        uri.scheme == null -> {
            val file = File(fontUrl)
            if (!file.isAbsolute) {
                return null
            }
            file.path
        }
        uri.scheme == "file" -> uri.path
        else -> null
    } ?: return null

    val file = File(path)
    val directory = if (file.isDirectory) file else file.parentFile ?: return null
    return directory.absolutePath
}

private fun JSObject.resolveTypstFontDirs(): List<String>? {
    if (!has("fontUrls") || isNull("fontUrls")) {
        return null
    }

    val fontUrls = optJSONArray("fontUrls")
        ?: throw IllegalArgumentException("Typst preview fontUrls must be an array of strings.")
    val fontDirs = buildList(fontUrls.length()) {
        repeat(fontUrls.length()) { index ->
            val fontUrl = fontUrls.optString(index, null)
                ?: throw IllegalArgumentException("Typst preview fontUrls must be strings.")
            val fontDir = resolveLocalFontDir(fontUrl)
                ?: throw IllegalArgumentException("Typst preview on mobile only supports local font file URLs or absolute font directories.")
            add(fontDir)
        }
    }
    return fontDirs.distinct()
}

@CapacitorPlugin(name = "Preview")
class PreviewPlugin : Plugin() {

    @PluginMethod
    fun renderMermaidSvg(call: PluginCall) {
        launch(Dispatchers.IO) {
            try {
                val code = call.getStringEnsure("code")
                val options = call.getObject("options")
                val svg = renderMermaidPreviewSvg(
                    code = code,
                    theme = options?.getOptionalString("theme"),
                    fontFamily = options?.getOptionalString("fontFamily"),
                    fontSize = options?.getOptionalDouble("fontSize"),
                )
                call.resolve(JSObject().apply {
                    put("svg", svg)
                })
            } catch (e: Exception) {
                Timber.e(e, "Failed to render Mermaid preview.")
                call.reject("Failed to render Mermaid preview.", null, e)
            }
        }
    }

    @PluginMethod
    fun renderTypstSvg(call: PluginCall) {
        launch(Dispatchers.IO) {
            try {
                val code = call.getStringEnsure("code")
                val options = call.getObject("options")
                val svg = renderTypstPreviewSvg(
                    code = code,
                    fontDirs = options?.resolveTypstFontDirs(),
                    cacheDir = context.cacheDir.absolutePath,
                )
                call.resolve(JSObject().apply {
                    put("svg", svg)
                })
            } catch (e: Exception) {
                Timber.e(e, "Failed to render Typst preview.")
                call.reject("Failed to render Typst preview.", null, e)
            }
        }
    }
}
